libcontainer/nsenter/nsexec.c

```diff
@@ -1,471 +1,878 @@
-	if (strcmp(initpipe, buf)) {
-		pr_perror("Unable to parse _LIBCONTAINER_INITPIPE");
-		exit(1);
+	if (write_file(policy, strlen(policy), "/proc/%d/setgroups", pid) < 0) {
+		/*
+		 * If the kernel is too old to support /proc/pid/setgroups,
+		 * open(2) or write(2) will return ENOENT. This is fine.
+		 */
+		if (errno != ENOENT)
+			bail("failed to write '%s' to /proc/%d/setgroups", policy, pid);
 	}
+}
 
-	return pipenum;
+static void update_uidmap(int pid, char *map, int map_len)
+{
+	if (map == NULL || map_len <= 0)
+		return;
+
+	if (write_file(map, map_len, "/proc/%d/uid_map", pid) < 0)
+		bail("failed to update /proc/%d/uid_map", pid);
 }
 
-// num_namespaces returns the number of additional namespaces to setns. The
-// argument is a comma-separated string of namespace paths.
-static int num_namespaces(char *nspaths)
+static void update_gidmap(int pid, char *map, int map_len)
 {
-	int i;
-	int size = 0;
+	if (map == NULL || map_len <= 0)
+		return;
 
-	for (i = 0; nspaths[i]; i++) {
-		if (nspaths[i] == ',') {
-			size += 1;
-		}
+	if (write_file(map, map_len, "/proc/%d/gid_map", pid) < 0)
+		bail("failed to update /proc/%d/gid_map", pid);
+}
+
+static char *trim(char *str, char *end, bool space) {
+	char *p = str;
+
+	while (p != end && isspace(*p) == space)
+		p++;
+	if (p == end)
+		return NULL;
+
+	return p;
+}
+
+/* Get the nth field from a space-separated map. */
+static int map_field(char *map, int map_len, int n)
+{
+	char *p = map;
+	char *end = map + map_len;
+	int i, value = 0;
+
+	/* Advance *p to the field. */
+	for (i = 0; i < n; i++) {
+		/* Trim the spaces. */
+		p = trim(p, end, true);
+		if (!p)
+			bail("unexpected end of map: '%s'", map);
+
+		/* Now we can skip over the actual field content. */
+		p = trim(p, end, false);
+		if (!p)
+			bail("unexpected end of map: '%s'", map);
 	}
 
-	return size + 1;
+	/* We don't need to trim spaces, strtol(3) handles it fine. */
+	value = strtol(p, &end, 10);
+	if (p == end || !isspace(*end))
+		bail("failed to parse field %d value: '%s'", n, map);
+
+	return value;
 }
 
-static uint32_t readint32(char *buf)
+/* A dummy function that just jumps to the given jumpval. */
+static int child_func(void *arg) __attribute__ ((noinline));
+static int child_func(void *arg)
 {
-    return *(uint32_t *)buf;
+	struct clone_t *ca = (struct clone_t *)arg;
+	longjmp(*ca->env, ca->jmpval);
 }
 
-static uint8_t readint8(char *buf)
+static int clone_parent(jmp_buf *env, int jmpval) __attribute__ ((noinline));
+static int clone_parent(jmp_buf *env, int jmpval)
 {
-    return *(uint8_t *)buf;
+	struct clone_t ca = {
+		.env    = env,
+		.jmpval = jmpval,
+	};
+
+	/*
+	 * Okay, so this is quite annoying.
+	 *
+	 * In order to make sure that deal with older kernels (when
+	 * CLONE_NEWUSER wasn't guaranteed to be done first if you specify
+	 * multiple namespaces in a clone(2) invocation) as well as with
+	 * certain usecases like rootless containers, we cannot just dump all
```

Comment:
can you describe the failure of rootless container usecase if we dont do CLONE_NEWUSER first ?

do we even want to support older kernel with malfunctioned clone 😢 ?